### DISCUSSION TITLE
Internalization and translation without sub- or domain routing.

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [vercel/next.js#56515] [Help]
### Summary

Hello and thanks for this unbelievably good meta-framework! 🚀

A key feature before I can fully move our company to nextJs is the translations. Currently the culture is chosen by the user at a settings page, then saved on the user settings. During initial loading we retrive the user and it's setting where the culutre is. and saves it in redux. On login page the user choose culture that is saved in a cookie.

I thought that translations in nextjs would be the least of my problem moving our application. But after a lot of reasearch I can't find a solution that do not utilze nextJs [Internationalization routing](https://nextjs.org/docs/app/building-your-application/routing/internationalization).

There are solutions that disguise the url culture path. But it's seems like really clunky and unnecessary solution imo (or in my case at least 😊).

We will use a state manager like Zustand or Jotai to manage the user and it's settings. And when the user changes the culture I will try to use server actions and revalidate the path to update language on the page. In otherwords I will always have access to correct culture in both Client and Server.

After this wall of text, my question is... do anyone know if there is a package out there that dont depend on the culture from the url? That only translates and keeps Client and Server at sync.

Maybe there Is a another good approach to this?

These are libraries I have been looking in to:

https://github.com/aralroca/next-translate

https://www.i18next.com/

https://next-international.vercel.app/

https://github.com/i18nexus/next-i18n-router

https://next-intl-docs.vercel.app/

Would love some thoughts arround this! 🙏

**Comment by J4v4Scr1pt** [comment]
I would love your thoughts about this but for anyone looking in to this, next-translate is truly awesome.
They also support queryParam as lang selector.
Just follow the normal install guide, but skip the add slug part [lang]. Then in the middleware add this:

```
export default async function middleware(request: NextRequest) {
	const locale = request.nextUrl.locale || i18n.defaultLocale; // Here you could use another saved user setting for locale
	request.nextUrl.searchParams.set('lang', locale);
	request.nextUrl.href = request.nextUrl.href.replace(`/${locale}`, '');
	return NextResponse.rewrite(request.nextUrl);
}
```
Thats it!

(Using App-folder)

**Comment by jalagania** [comment]
Thanks for this! Did you find any better solution? Do you still use next-translate?

You can have a look at this post, though I haven't tried it myself yet. 
https://carlogino.com/blog/nextjs-app-dir-i18n-cookie
